org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/model/TestSuiteElement.java

```diff
@@ -47,7 +48,8 @@ public String getSuiteTypeName() {
 
 	@Override
 	public ITestElement[] getChildren() {
```

Comment:
this is not enough. instead, copyonwrite list should be used

Reply:
How is it not enough, when the code on `master` already does the same and doesn't run into a CME? Also a copy-on-write list will make the `add()` method explode in complexity.

I've looked at `java.util.ArrayList.toArray(T[])`, the implementation refers to the array pointer when copying elements from it:

```
    public <T> T[] toArray(T[] a) {
        if (a.length < size)
            // Make a new array of a's runtime type, but my contents:
            return (T[]) Arrays.copyOf(elementData, size, a.getClass());
        System.arraycopy(elementData, 0, a, 0, size);
        if (a.length > size)
            a[size] = null;
        return a;
    }
```

I.e. when the `ArrayList` gets expnanded/shrinked when adding/removing new elements, a concurrent `toArray()` method call is still using the old array pointer (and so no CME occurs).

From my POV the other alternative is to add some synchronization here. Right now the `getChildren()` method costs twice as much as before (two copies instead of one, plus some effort for filtering). Synchronization is IMO more complicated and the change itself is more complicated to do.

I ran `org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/TestAll.java`, it takes about 17 minutes with and without the change. I don't see a difference, I can point you to the screen recordings of the runs (two recordings, with and without the change respectively).